Add Kubernetes events watcher websocket endpoint

jiekang opened the pull request:
This addresses the 'events' side of https://github.com/openshiftio/openshift.io/issues/2380

The endpoint `api/deployments/environments/{environmentId}/events/watch` is added that allows websocket clients to connect and receive the Kubernetes Events ListAndWatch results.

This is a WIP PR and contains commits and files that need to be cleaned before merging. These are:

Commit: 
temp(deployments): set localhost
fix(deployments): add readme for websockets prototype

Files:
examples/client.go

jiekang commented:
@stooke @ebaron could you take a look at this and see if there are any flaws in my implementation?

As well, when running this locally and connecting to it via my client, the server logs have output:

```
10:28:26 app         | E0523 10:28:26.483742   29272 streamwatcher.go:109] Unable to decode an event from the watch stream: net/http: request canceled (Client.Timeout exceeded while reading body)
```
every 30 seconds or so. This code is in Kubernetes: https://github.com/kubernetes/apimachinery/blob/master/pkg/watch/streamwatcher.go

I believe the timeout from the kubernetes server is intended [1] but 30 seconds is quite short and the messages fill the log. I'm not sure how I can mitigate this. Thoughts? Am I missing something?

[1]
https://github.com/kubernetes/apimachinery/issues/41

jiekang commented:
Adding tests is also on my to-do list...

ebaron commented on controller/deployments.go:
```diff
@@ -381,6 +384,67 @@ func (c *DeploymentsController) ShowSpaceEnvironments(ctx *app.ShowSpaceEnvironm
+			log.Error(ctx, map[string]interface{}{
+				"err": err,
+			}, "error accessing Auth server")
+			return
```
Is there some way we can inform the client of errors that may occur here and throughout this method?

jiekang replied:
Ah good point. At the moment this endpoint is just repeating the results of the Kubernetes watch and listen to the client. It can also be made to send useful error messages to the client via the websocket.

As well, instead of repeating the results raw, I was thinking about whether or not we want to massage the data a little to tailor it to the UX needs, like how our other deployments endpoints are. Thoughts?

ebaron replied:
Tailoring the responses to the frontend makes sense to me. As you said, we already do it elsewhere.

ebaron commented on controller/deployments.go:
```diff
@@ -381,6 +384,67 @@ func (c *DeploymentsController) ShowSpaceEnvironments(ctx *app.ShowSpaceEnvironm
+				if err != cache.FIFOClosedError {
+					log.Error(ctx, map[string]interface{}{
+						"err": err,
+					}, "error receiving events: "+time.Now().String())
```
I think the timestamp here is redundant, the WIT logger adds a timestamp to all messages.

ebaron commented on controller/deployments.go:
```diff
@@ -381,6 +384,67 @@ func (c *DeploymentsController) ShowSpaceEnvironments(ctx *app.ShowSpaceEnvironm
+			if err != nil {
+				log.Error(ctx, map[string]interface{}{
+					"err": err,
+				}, "error sending events: "+time.Now().String())
```
Same as above.

stooke commented:
Has this been tested against the f8-proxy, or directly to k8s, or both?

jiekang commented:
I've only tested this in a locallly running instance targeting production. It uses the same underlying system as the other APIs to connect to Kubernetes so I am expecting it to work since the other APIs work. I can look further if you have more directions on how to test against f8-proxy or directly to k8s.

jiekang commented:
I've added some tests for the kubeclient. I'm looking at tests for the deployments controller but the goa test generator doesn't create helpers for endpoints with ws/wss scheme so it may be a lot more complicated.

I've used gops to monitor the goroutines. On my local instance, each connection spawns 8 goroutines and closing the connection does clean them up properly. Does 8 routines for one connection seem a bit high?

@ebaron Could you take a look at the tests I've added to `kubernetes/deployments_kubeclient_blackbox_test.go` I would not be surprised if there's a better way. I have very little experience with go...

ebaron commented:
@jiekang The communication with OpenShift still uses HTTP and not WebSockets, doesn't it? If so, perhaps we could use go-vcr to play back test responses instead of re-adding this manual mocking code.

jiekang commented:
I can look into that; I wasn't sure how exactly the internals of the kubernetes code worked with http so I chose to mock it instead. I feel like it becomes less of a blackbox that way though to be fair I did need to know some internals to write these tests too hahah.

jiekang commented:
After investigation, the watch API implementation uses HTTP streaming via `Transfer-Encoding: chunked` and I don't think go-vcr can be used to simulate this.

For simple HTTP requests, it's unlikely the internal implementation will switch to a completely different technology and I understand how it reduces the amount of code needed, however I still think that would make these unit tests too dependent on the API implementation details. For example, if the kubernetes client switched internally from HTTP streaming to websockets, users would likely not have to change anything, as the streaming mechanism from the results is abstracted by the API, but these go-vcr tests would no longer be accurate.

ebaron commented:
@jiekang In the absence of another way to test this, I'm fine with keeping the approach in this PR. With the fix for https://github.com/openshiftio/openshift.io/issues/3111, we should be able to remove the Config Map boilerplate code at least.

jiekang commented:
@ebaron I finished the unit tests for the deployments controller; `deployments_blackbox_test.go`. Could you review the code I wrote there when you have time?

My next step is to modify the data returned to remove all the useless things, and then update the tests to use mock data that matches this.

ebaron commented on controller/deployments_blackbox_test.go:
```diff
@@ -587,3 +595,152 @@ func createOSIOClientMock(t minimock.Tester, spaceName string) *testcontroller.O
+			conn.Read(buf)
+			frameLength := int(buf[1])
+			startPos := 2
+			endPos := startPos + frameLength
```
It's a shame we have to handle WebSocket frames manually, but I don't see a way around it unless we somehow return a websocket.Conn instead of a net.Conn from WatchEnvironmentEventsDeploymentsOK. Maybe add a link to the RFC section that details the frame encoding [1]? I found this helpful to understand what's going on here.

[1] https://tools.ietf.org/html/rfc6455#section-5.2

jiekang commented:
@ebaron I tried to address your review, thanks! I've made some significant additions to transform the data (really just reduce the fields returned), discussed in https://github.com/openshiftio/openshift.io/issues/2528. It's mostly done now. Could you take a look when you have time? Thanks a lot!

ebaron commented on controller/deployments.go:
```diff
@@ -441,17 +441,74 @@ func (c *DeploymentsController) WatchEnvironmentEventsWSHandler(ctx *app.WatchEn
+	// Null for lists.
+	// More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata
+	// +optional
+	CreationTimestamp metaV1.Time `json:"creationTimestamp,omitempty" protobuf:"bytes,8,opt,name=creationTimestamp"`
```
Perhaps you can use a Field Selector to limit the included fields, instead of having to transform the data yourself? https://docs.openshift.com/online/rest_api/api/v1.Event.html#Get-api-v1-watch-namespaces-namespace-events

jiekang replied:
My understanding is that FieldSelector limits the set of objects returned based by some conditions on the object fields, e.g. return all events with field `X` with value `Y`. Here we're modifying the returned object to have a subset of fields with the set of objects being unchanged. Does that make sense?

ebaron replied:
You're correct. The documentation for field selectors is lacking...

ebaron commented on controller/deployments_blackbox_test.go:
```diff
@@ -736,21 +748,47 @@ func TestWatchEnvironmentEvents(t *testing.T) {
 }
 
+func transformItem(item interface{}) (*controller.DeploymentsEvent, error) {
+	if event, ok := item.(*v1.Event); ok {
```
Item is guaranteed to be a `v1.Event` isn't it? We could get rid of the type-assertion and error return value here.

ebaron commented on controller/deployments.go:
```diff
@@ -404,6 +409,138 @@ func (c *DeploymentsController) ShowAllEnvironments(ctx *app.ShowAllEnvironments
+}
+
+func transformItem(item interface{}) (*DeploymentsEvent, error) {
+	if event, ok := item.(*v1.Event); ok {
```
Item is guaranteed to be a `v1.Event` isn't it? We could get rid of the type-assertion and error return value here.

jiekang replied:
I'm not sure the best way to go about this. The input for this comes from the Kubernetes Store that is returned from
```
store, stopWs := kc.WatchEventsInNamespace(ctx.EnvName)
```

The definition for `FIFO` `pop` is `func (f *FIFO) Pop(process PopProcessFunc) (interface{}, error) {`

Is there a way I can rework it to give me a `*v1.Event` so I don't need to to the type transformation? Actually, I guess I can do it earlier so `transformItem` takes a `*v1.Event`. That will be less confusing.

ebaron replied:
Ah, sorry this comment was meant for the `transformItem` in the unit tests, where the argument passed is of Event type already.

jiekang replied:
Oh, no worries. I've changed the one in the unit tests. I've also reworked the one in the controller too so the type transform is earlier.

jiekang commented:
@ebaron so it looks like I hit a snag with using this on the client side. The javascript websocket API that is used for browser-based clients does **not** support setting headers for the HTTP upgrade GET request. That means I can't set the `Authorization` header that is used by the WIT ws API to authenticate and authorize the connection. 

Do you know if it's possible to set the goa design to give us code that will check a query parameter like `access_token` for the JWT token instead of the `Authorization` header, just for this endpoint? I'm a bit skeptical...

jiekang commented:
After some digging I tried to add a JWT auth mechanism that uses the query parameter, however it turns out the functionality is not yet implemented [1]. I guess I could take a look at opening a PR for them...

[1]
https://github.com/goadesign/goa/blob/master/middleware/security/jwt/jwt.go#L67

jiekang commented:
@aslakknutsen I ran into a possible showstopper for using goa and websockets where the client side browser-based application doesn't have support for an upgrade GET request with an auth header. That is, the javascript API for websockets [0] does not allow for setting header fields for the upgrade GET request.

One alternative is sending the token through as a query parameter, though it is not implemented yet in Goa [1]. I would be able to send a PR for that, but query parameter may not be a safe option anyways.

Do you have any thoughts on using the query parameter and what should be done in general? Maybe dropping websockets entirely and using HTTP streaming instead? 

[0] https://developer.mozilla.org/en-US/docs/Web/API/WebSocket
[1] https://github.com/goadesign/goa/blob/master/middleware/security/jwt/jwt.go#L67

jiekang commented:
For what it's worth, the existing pipelines code opens websockets via request like:

```
wss://f8osoproxy-test-dsaas-production.09b5.dsaas.openshiftapps.com/oapi/v1/namespaces/jkang/buildconfigs?watch=true&access_token=<token>
```

This uses a query parameter `access_token` for auth.

jiekang commented:
@ebaron Hey! I've updated the code to support JWT auth via query parameter. When available, could you give it another review? Thanks!

The only thing I plan to do is update the commit messages to include the tracking issue, and maybe add comments to the file itself (jwt_middleware.go) to also include the tracking issue.

ebaron commented on goamiddleware/jwt_middleware.go:
```diff
@@ -0,0 +1,284 @@
+
+const (
+	jwtKey contextKey = iota + 1
+)
```
It looks like this type and constant aren't used.

ebaron commented on goamiddleware/jwt_middleware.go:
```diff
@@ -0,0 +1,284 @@
+		}
+	}
+	return
+}
```
Is it possible to use the versions of these functions in jwt_token_context.go? Or move any shared functions to a common utility file?

ebaron replied:
Shared functions: parse/validate{RSA,ECDSA,HMAC}Keys, partitionKeys

ebaron commented on goamiddleware/jwt_middleware.go:
```diff
@@ -0,0 +1,284 @@
+
+// ErrJWTError is the error returned by this middleware when any sort of validation or assertion
+// fails during processing.
+var ErrJWTError = goa.NewErrorClass("jwt_security_error", 401)
```
We could use goajwt.ErrJWTError instead of redeclaring it.

stooke replied:
"could" -> "should"

stooke commented:
To test using the production proxy, set
osoproxy.url : https://f8osoproxy-test-dsaas-production.09b5.dsaas.openshiftapps.com
in your config.yaml.  I'm not sure if there's more you have to do, like set "FABRIC8_WIT_API_URL=https://api.openshift.io F8_CONFIG_FILE_PATH=your-config.yaml" in your environment.

stooke commented:
Have you tested restart (or don't care) about the 30 minute k8s timeout?

ebaron commented on log/log.go:
```diff
@@ -141,6 +141,14 @@ func Error(ctx context.Context, fields map[string]interface{}, format string, ar
+							params[string(k)] = v
+						}
+					}
 					entry = entry.WithField("req_params", req.Params)
```
This should be `params` and not `req.Params`.

jiekang commented:
@stooke I believe the go client's Reflector internally handles the timeout from the kubernetes apiserver. I'll do some long lived testing to verify this.

jiekang commented:
@stooke I modified my config via
```
diff --git a/jkconfig.yaml b/jkconfig.yaml
index 03c67db..63b2477 100644
--- a/jkconfig.yaml
+++ b/jkconfig.yaml
@@ -1 +1,2 @@
 auth.url : https://auth.openshift.io
+osoproxy.url : https://f8osoproxy-test-dsaas-production.09b5.dsaas.openshiftapps.com
```

and ran a local instance. I was able to connect properly and see events. Is there anything else I should validate?
